### PR TITLE
[ValueTracking] Expand cmpExcludesZero to optionally work with non-constant RHS

### PR DESCRIPTION
**Pull request by goldsteinn**
- [ValueTracking] Add tests for proving phi-nonzero based on ICmp without constants; NFC
- [ValueTracking] Expand `cmpExcludesZero` to optionally work with non-constant RHS

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
Could you please provide the original motivating case for this? I think this goes beyond what is reasonable to handle in ValueTracking, and it would be best to handle your use case via LVI+CVP instead, if this is possible.

**Comment by goldsteinn** [comment]
> Could you please provide the original motivating case for this? I think this goes beyond what is reasonable to handle in ValueTracking, and it would be best to handle your use case via LVI+CVP instead, if this is possible.

No strong motivating force. Guess the thinking was this is still basically a cheap check:
https://llvm-compile-time-tracker.com/compare.php?from=0ec39fb48e9dc4aad5f20bc50d44947ac004dd34&to=9031c8347dd041a4f34d8499d9e842ca9c33fd27&stat=instructions:u
And just improves coverage.
Can try adding to LVI, however.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -562,25 +564,37 @@ static bool cmpExcludesZero(CmpInst::Predicate Pred, const Value *RHS) {
     return match(RHS, m_Zero());
 
   // All other predicates - rely on generic ConstantRange handling.
+
```
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:567]: Drop the newline here.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 2]
```diff
@@ -552,7 +552,9 @@ bool llvm::isValidAssumeForContext(const Instruction *Inv,
 // example Pred=EQ, RHS=isKnownNonZero. cmpExcludesZero is called in loops
 // so the extra compile time may not be worth it, but possibly a second API
 // should be created for use outside of loops.
-static bool cmpExcludesZero(CmpInst::Predicate Pred, const Value *RHS) {
+static bool cmpExcludesZero(CmpInst::Predicate Pred, Value *RHS,
+                            const SimplifyQuery *Q = nullptr,
```
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:556]:
```suggestion
                            const SimplifyQuery &Q,
```
It is strange to me.  I have checked all callers of `cmpExcludeZero` and found that `SimplifyQuery` is always available.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 3]
```diff
@@ -2666,15 +2680,21 @@ static bool isKnownNonZeroFromOperator(const Operator *I,
 
       // The condition of the select dominates the true/false arm. Check if the
       // condition implies that a given arm is non-zero.
-      Value *X;
+      Value *X, *Y;
       CmpInst::Predicate Pred;
-      if (!match(I->getOperand(0), m_c_ICmp(Pred, m_Specific(Op), m_Value(X))))
+      if (!match(I->getOperand(0), m_ICmp(Pred, m_Value(X), m_Value(Y))))
```
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:2685]: It seems like a NFC change. Does this change improve matching performance?

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 4]
```diff
@@ -2696,18 +2716,25 @@ static bool isKnownNonZeroFromOperator(const Operator *I,
       RecQ.CxtI = PN->getIncomingBlock(U)->getTerminator();
       // Check if the branch on the phi excludes zero.
       ICmpInst::Predicate Pred;
-      Value *X;
+      Value *X, *Y;
       BasicBlock *TrueSucc, *FalseSucc;
       if (match(RecQ.CxtI,
-                m_Br(m_c_ICmp(Pred, m_Specific(U.get()), m_Value(X)),
+                m_Br(m_ICmp(Pred, m_Value(X), m_Value(Y)),
```
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:2722]: So is this one.

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 3 above)
Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:2685]: Its necessary now that we can use non-constants. I'll seperate that to an NFC and rebase this.

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:2685]: Will probably abandon this one per nikic's comments, but will update the matching logics

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 2 above)
Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:556]: Its intentional so we can still have the const-only version. the perf cost of doing recursive CR call in the assume loop for example would be significant.

**Pull request closed without merging**